text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,325 @@
+
+Many people have already invested a lot of effort into writing code against the CDK Pipelines
+library. The changes we are planning mostly affect the internals of the library, and we want to add
+some slight enhanced expressiveness and customizability to the API.
+
+With a little additional effort, we can make the new API additive and have the old one continue to work,
+giving people the opportunity to switch over to the new API at their own pace.
+
+### How will we stay backwards compatible?
+
+Right now, the parts that are too limiting are the parts that are directly re-exposed from the `aws-codepipeline`
+library. For example, the fact that the synth needs to be a `codepipeline.IAction`.
+
+What we'll do is type testing to switch between old and new behavior. Conceptually:
+
+```ts
+interface IModernSynthAction implements codepipeline.IAction {
+  // Or whatever the signature ends up being
+  public addToPipelineModern(pipeline: CdkPipeline): AddResult;
+}
+
+if (isModernSynthAction(props.synthAction)) {
+  // New behavior
+  props.synthAction.addToPipelineModern(this);
+} else {
+  // Wrapper class for legacy behavior
+  new LegacySynthAction(props.synthAction).addToPipelineModern(this);
+}
+```
+
+### How do we render deployments out across backends?
+
+In the middle layer, a stack deployments gets represented as nested state machines,
+like this:
+
+```
+┌───────────────────────────────────────────────────────────────┐
+│                          DeployStack                          │
+│                                                               │
+│     ┌────────────────────┐         ┌────────────────────┐     │
+│     │                    │         │                    │     │
+│     │  CreateChangeSet   │────────▶│  ExecuteChangeSet  │     │
+│     │                    │         │                    │     │
+│     └────────────────────┘         └────────────────────┘     │
+│                                                               │
+└───────────────────────────────────────────────────────────────┘
+```
+
+A backend rendered has the choice to either:
+
+* Recognize a `DeployStack` state and render that out to a `cdk deploy` command (or something
+  else appropriate); or
+* Recognize the individual `CreateChangeSet` and `ExecuteChangeSet` states and render those
+  out either to CodePipeline actions or `aws cloudformation change-change-set` CLI commands,
+  as may be appropriate.
+
+### How do approval workflows work?
+
+Every `Approval` step will be handed a reference to the workflow of the application
+being deployed, as well as to an empty "approval" workflow. It can then choose to
+add new actions and dependencies to any of those workflows.
+
+In the most common case, it will add things like "shell script" actions to the approval workflow; but it might also
+choose to add actions before "Deploy" (`Approval.securityChanges()`), or it might choose to add actions *in between*
+`CreateChangeSet` and `DeployChangeSet` pairs.
```

Comment:
:-) see my comment above... If a backend ignores CreateChangeSet and DeployChangeSet how will an approval workflow that is executed between them work?

Reply:
In that case, it will throw an error saying "this kind of validation can not be used with this type of backend". 

Which again is fine, right. Other types of validations WILL work, just not this particular one.

Reply:
Who will throw this error? The backend? How will it even know that this validation is defined if it ignores the "Create/DeployChangeSet" actions?